eos_data_distribution/avahi_monitor.py

```diff
@@ -76,26 +112,131 @@ def __init__(self):
+                     name, h_type, type, domain)
+
+        self.refresh_network()
+
+        if address in self.ips:
+            return
+
         faceUri = face_uri_from_triplet(type, host, port)
-        check_call(["nfdc", "add-nexthop",
-                    "-c", str(defaults.RouteCost.LOCAL_NETWORK),
-                    str(SUBSCRIPTIONS_SOMA), faceUri])
-        self._registry[build_registry_key(name, type, domain)] = faceUri
+        key = build_registry_key(name, type, domain)
+        self._registry[key] = faceUri
+
+        if address in self.gateways:
+            self._ndn_gateways[key] = address
+
+        self.update_routes()
 
     def service_removed_cb(self, sda, interface, protocol, name, type, domain, flags):
-        ifname = sda.siocgifname(interface)
-        print "Disappeared Service '%s' of type '%s' in domain '%s' on %s.%i." % (name, type, domain, ifname, protocol)
-        faceUri = self._registry[build_registry_key(name, type, domain)]
-        check_call(
-            ["nfdc", "remove-nexthop", str(SUBSCRIPTIONS_SOMA), faceUri])
+        logger.debug(
+            "Disappeared Service '%s' of type '%s' in domain '%s'",
+            name, type, domain)
+
+        self.refresh_network()
+
+        key = build_registry_key(name, type, domain)
+        try:
+            del self._ndn_gateways[key]
+        except KeyError:
+            pass
+
+        faceUri = self._registry[key]
+        if self.remove_nexthop(faceUri)
+        del self._registry[key]
+
+        prev_routed = self._routed
+        self.update_routes()
+
+    def add_nexthop(self, faceUri):
+        if len(self._peers) > MAX_PEERS:
+            logger.debug(
+                "refusing to add %s as we'd go over %d peers", faceUri, MAX_PEERS)
+            return
+
+        if self.check_call(["nfdc", "add-nexthop",
+                            "-c", str(defaults.RouteCost.LOCAL_NETWORK),
+                            str(SUBSCRIPTIONS_SOMA), faceUri]):
+            self._peers.add(faceUri)
+
+    def remove_nexthop(self, faceUri):
+        try:
+            self._peers.remove(faceUri)
+            self.check_call(["nfdc", "remove-nexthop",
+                             str(SUBSCRIPTIONS_SOMA), faceUri])
+        except KeyError:
+            pass
+
+    def start_edge(self):
+        return self.check_call(["systemctl", "start", "edd-soma-subscriptions-producer"])
+
+    def stop_edge(self):
+        return self.check_call(["systemctl", "stop", "edd-soma-subscriptions-producer"])
+
+    def get_gateways(self):
+        gateways = netifaces.gateways()
+        del gateways['default']
+        return [g[0] for g in flatten(gateways.values())]
+
+    def refresh_network(self, *args, **kwargs):
+        self.gateways = self.get_gateways()
+        self.ips = flatten(
+            map(lambda a: [x['addr'] for x in a], [flatten(netifaces.ifaddresses(i).values()) for i in netifaces.interfaces()]))
```

Comment:
there's a lot going on here, IMO it'd be much clearer if split up over a few lines